Add a CONTRIBUTING.md for the distributed package.

pritamdamania87 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#44224 Add a CONTRIBUTING.md for the distributed package.**

The purpose of this file is to help developers on PT distributed get
upto speed on the code structure and layout for PT Distributed.

Differential Revision: [D23548377](https://our.internmc.facebook.com/intern/diff/D23548377/)

pritamdamania committed:
Add a CONTRIBUTING.md for the distributed package.

The purpose of this file is to help developers on PT distributed get
upto speed on the code structure and layout for PT Distributed.

Differential Revision: [D23548377](https://our.internmc.facebook.com/intern/diff/D23548377/)

[ghstack-poisoned]

pritamdamania committed:
Update on "Add a CONTRIBUTING.md for the distributed package."


The purpose of this file is to help developers on PT distributed get
upto speed on the code structure and layout for PT Distributed.

Differential Revision: [D23548377](https://our.internmc.facebook.com/intern/diff/D23548377/)

[ghstack-poisoned]

rohan-varma commented:
This is great! Shall we include some information on Distributed specific development tips, such as how to run the tests for RPC (maybe the various backends as well), DDP, and collective comm APIs?

rohan-varma commented on torch/distributed/CONTRIBUTING.md:
```diff
@@ -0,0 +1,69 @@
+
+### Distributed Data Parallel
+
+DDP is implemented as a module in [distributed.py](../nn/parallel/distributed.py) with some of the core functions implemented in [reducer.cpp](../csrc/distributed/c10d/reducer.cpp) and [comm.cpp](../csrc/distributed/c10d/reducer.cpp). Gradients synchronizations occur in backward pass, triggered as autograd hooks.
```
Should we link to DDP design (https://pytorch.org/docs/stable/notes/ddp.html) and maybe do the same for RPC / Dist Autograd? I guess we may not need to since these links are available from the dist_overview page.

rohan-varma commented on torch/distributed/CONTRIBUTING.md:
```diff
@@ -0,0 +1,69 @@
+
+### RPC Agents
+
+The core C++ interface of the RPC framework can be found in [rpc_agent.h](../csrc/distributed/rpc/rpc_agent.h) and the TensorPipe and ProcessGroup implementations can be found at [process_group_agent.h](../csrc/distributed/rpc/process_group_agent.h) and [tensorpipe_agent.h](../csrc/distributed/rpc/tensorpipe_agent.h) respectively.
```
Nit: should we specify Process group Gloo backend instead of just Process Group backend?

pritamdamania committed:
Update on "Add a CONTRIBUTING.md for the distributed package."


The purpose of this file is to help developers on PT distributed get
upto speed on the code structure and layout for PT Distributed.

Differential Revision: [D23548377](https://our.internmc.facebook.com/intern/diff/D23548377/)

[ghstack-poisoned]
